### DISCUSSION TITLE
Calling any C++ developers to help with an upstream bug in Tracktion Engine (No JUCE/Tracktion experience needed)

### DISCUSSION THREAD

**Discussion by stonepreston** [FundamentalFrequency/.github#12] [Announcements]
Hey folks,

There is currently a problem in the upstream tracktion engine preventing the engine from being compiled using newer compilers like Clang-14 and GCC 11.2.0. Ive opened an issue the in LMN-3 DAW [here](https://github.com/FundamentalFrequency/LMN-3-DAW/issues/45). There is a link to a forum post in there for more details.

This seems to be a compiler issue with the allocator being used in line 161 [here](https://github.com/Tracktion/tracktion_engine/blob/48e837ae5e248888dc587458ca5926da4ad526d4/modules/tracktion_graph/utilities/tracktion_Allocation.test.cpp#L161). If you comment out 161 it compiles but thats obviously not a fix. 

I believe this bug could be resolved by someone with some C++ experience, and you dont need previous JUCE experience really as it seems to be purely C++ related. 

To replicate the issue (ensure you install the [build dependencies](https://github.com/FundamentalFrequency/LMN-3-DAW#dependencies) you will need to use a newer compiler (Ive tried with both Clang-14 and GCC 11.2.0, both of which are in the repositories of Ubuntu 22.04)

```
git clone --recurse-submodules https://github.com/Tracktion/tracktion_engine.git
cd tracktion_engine
git checkout develop
cd examples/DemoRunner/
cmake -B build
cmake --build build
```

The error spit out is:
```
home/stone/development/tracktion_engine/modules/tracktion_graph/utilities/tracktion_Allocation.test.cpp:161:54:   required from here
/usr/include/c++/11/bits/stl_vector.h:1483:9: error: no match for ‘operator==’ (operand types are ‘std::_Vector_base<int, rpallocator<int> >::_Tp_alloc_type’ {aka ‘std::allocator_traits<rpallocator<int> >::rebind_alloc<int>’} and ‘std::_Vector_base<int, rpallocator<int> >::_Tp_alloc_type’ {aka ‘std::allocator_traits<rpallocator<int> >::rebind_alloc<int>’})
 1483 |         __glibcxx_assert(_Alloc_traits::propagate_on_container_swap::value
      |         ^~~~~~~~~~~~~~~~
```

This technically is not a show stopper, but Id like to get the engine building with these newer compilers as they are now the defaults in a lot of distributions. 

I will have some time tomorrow to dig in as well, but could use any help I can get to tackle this.

**Comment by stonepreston** [comment]
I have found a fix for the issue and created a [PR](https://github.com/Tracktion/tracktion_engine/pull/120) into tracktion engine.

**Reply by stonepreston** [reply]
Fix is now on develop

**Comment by abeck99** [comment]
Hi - I have extensive 25 years c++ experience and have a deep knowledge of clang/multi platform since I've been making a commercial llvm based audio programming language for the last 4 years as well as some experience with ATMega chips. I love this project and the amount of work you've done so far really sets the stage for OS development. If you need any help debugging low-level C++ issues I'd be happy to help. Amazing work man
